spec/services/external_upload_manager_spec.rb

```diff
@@ -0,0 +1,220 @@
+      let(:metadata_headers) { { "x-amz-meta-sha1-checksum" => client_sha1 } }
+
+      context "when the downloaded file sha1 does not match the client sha1" do
+        let(:client_sha1) { "blahblah" }
+
+        it "raises an error and marks upload as failed" do
+          expect { subject.promote_to_upload! }.to raise_error(ExternalUploadManager::ChecksumMismatchError)
+          expect(external_upload_stub.reload.status).to eq(ExternalUploadStub.statuses[:failed])
+        end
+      end
+    end
+  end
+
+  context "when stubbed upload is > DOWNLOAD_LIMIT (too big to download, generate a fake sha)" do
+    let(:object_size) { 200.megabytes }
+    let(:object_file) { pdf_file }
+    let!(:external_upload_stub) { Fabricate(:attachment_external_upload_stub, created_by: user) }
+
+    before do
+      UploadCreator.any_instance.stubs(:generate_fake_sha1_hash).returns("testbc60eb18e8f974cbfae8bb0f069c3a311024")
+    end
+
+    it "does not try and download the file" do
+      FileHelper.expects(:download).never
+      subject.promote_to_upload!
+    end
+
+    it "generates a fake sha for the upload record" do
+      upload = subject.promote_to_upload!
+      expect(upload.sha1).not_to eq(sha1)
+      expect(upload.original_sha1).to eq(nil)
+      expect(upload.filesize).to eq(object_size)
+    end
+
+    it "marks the stub as uploaded" do
+      subject.promote_to_upload!
+      expect(external_upload_stub.reload.status).to eq(ExternalUploadStub.statuses[:uploaded])
+    end
+
+    it "copies the stubbed upload on S3 to its new destination and deletes it" do
+      upload = subject.promote_to_upload!
+      expect(WebMock).to have_requested(
+        :put,
+        "#{upload_base_url}/original/1X/#{upload.sha1}.pdf"
+      ).with(headers: { 'X-Amz-Copy-Source' => "#{SiteSetting.s3_upload_bucket}/#{external_upload_stub.key}" })
+      expect(WebMock).to have_requested(
+        :delete, "#{upload_base_url}/#{external_upload_stub.key}"
+      )
+    end
+  end
+
+  def stub_head_object
+    stub_request(
+      :head,
+      "#{upload_base_url}/#{external_upload_stub.key}"
+    ).to_return(
+      status: 200,
+      headers: {
+        ETag: etag,
+        "Content-Length" => object_size,
+        "Content-Type" => "image/png",
+      }.merge(metadata_headers)
+    )
+  end
+
+  def stub_download_object_filehelper
+    signed_url = Discourse.store.signed_url_for_path(external_upload_stub.key)
+    uri = URI.parse(signed_url)
+    signed_url = uri.to_s.gsub(uri.query, "")
+    stub_request(:get, signed_url).with(query: hash_including({})).to_return(
+      status: 200,
+      body: object_file.read
+    )
+  end
+
+  def stub_copy_object
+    copy_object_result = <<~BODY
+<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n
+<CopyObjectResult
+  xmlns=\"http://s3.amazonaws.com/doc/2006-03-01/\">
+  <LastModified>2021-07-19T04:10:41.000Z</LastModified>
+  <ETag>&quot;#{etag}&quot;</ETag>
+</CopyObjectResult>
+    BODY
```

Comment:
```suggestion
    copy_object_result = <<~BODY
    <?xml version=\"1.0\" encoding=\"UTF-8\"?>\n
    <CopyObjectResult
      xmlns=\"http://s3.amazonaws.com/doc/2006-03-01/\">
      <LastModified>2021-07-19T04:10:41.000Z</LastModified>
      <ETag>&quot;#{etag}&quot;</ETag>
    </CopyObjectResult>
    BODY
```

I think this should work since we're using squishy heredoc.